storage/src/table/storage.rs

```diff
@@ -0,0 +1,475 @@
+    const SIZE: usize = TABLE_ENTRY_SIZE;
+}
+
+impl CodecWrite for TableEntry {
+    fn write(&self, buf: &mut impl BufMut) {
+        self.epoch.write(buf);
+        self.section.write(buf);
+        self.offset.write(buf);
+        self.crc.write(buf);
+    }
+}
+
+impl Read for TableEntry {
+    type Cfg = ();
+
+    fn read_cfg(buf: &mut impl Buf, _: &Self::Cfg) -> Result<Self, commonware_codec::Error> {
+        let epoch = u64::read(buf)?;
+        let section = u64::read(buf)?;
+        let offset = u32::read(buf)?;
+        let crc = u32::read(buf)?;
+
+        Ok(Self {
+            epoch,
+            section,
+            offset,
+            crc,
+        })
+    }
+}
+
+/// Record stored in the journal for linked list entries.
+struct JournalEntry<K: Array, V: Codec> {
+    key: K,
+    value: V,
+
+    next: Option<(u64, u32)>,
+}
+
+impl<K: Array, V: Codec> JournalEntry<K, V> {
+    /// Create a new `JournalEntry`.
+    fn new(key: K, value: V, next: Option<(u64, u32)>) -> Self {
+        Self { key, value, next }
+    }
+}
+
+impl<K: Array, V: Codec> CodecWrite for JournalEntry<K, V> {
+    fn write(&self, buf: &mut impl BufMut) {
+        self.key.write(buf);
+        self.value.write(buf);
+        self.next.write(buf);
+    }
+}
+
+impl<K: Array, V: Codec> Read for JournalEntry<K, V> {
+    type Cfg = V::Cfg;
+
+    fn read_cfg(buf: &mut impl Buf, cfg: &Self::Cfg) -> Result<Self, commonware_codec::Error> {
+        let key = K::read(buf)?;
+        let value = V::read_cfg(buf, cfg)?;
+        let next = Option::<(u64, u32)>::read_cfg(buf, &((), ()))?;
+
+        Ok(Self { key, value, next })
+    }
+}
+
+impl<K: Array, V: Codec> EncodeSize for JournalEntry<K, V> {
+    fn encode_size(&self) -> usize {
+        K::SIZE + self.value.encode_size() + self.next.encode_size()
+    }
+}
+
+/// Implementation of immutable key-value storage.
+pub struct Table<E: Storage + Metrics + Clock, K: Array, V: Codec> {
+    // Context for storage operations
+    context: E,
+
+    // Table size
+    table_partition: String,
+    table_size: u32,
+
+    // Table blob that maps hash values to journal locations (journal_id, offset)
+    table: E::Blob,
+
+    // Variable journal for storing entries
+    journal: Journal<E, JournalEntry<K, V>>,
+
+    // Target size of each journal
+    target_journal_size: u64,
+
+    // Current section for new writes
+    current_section: u64,
+    next_epoch: u64,
+
+    // Metrics
+    puts: Counter,
+    gets: Counter,
+
+    // Pending table updates to be written on sync (table_index -> (section, offset))
+    modified_sections: BTreeSet<u64>,
+
+    // Phantom data to satisfy the compiler about generic types
+    _phantom: PhantomData<(K, V)>,
+}
+
+impl<E: Storage + Metrics + Clock, K: Array, V: Codec> Table<E, K, V> {
+    /// Initialize a new [Table] instance.
+    pub async fn init(
+        context: E,
+        config: Config<V::Cfg>,
+        epoch: u64,
+        size: (u64, u64),
```

Comment:
This is the most straightforward way to let a "root metadata" drive the `Table` state machine forward.